ee/server/service/hosts.go

```diff
@@ -30,3 +35,249 @@ func (svc *Service) OSVersion(ctx context.Context, osID uint, teamID *uint, incl
+	case lockWipe.IsPendingUnlock():
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Host has pending unlock request. Host cannot be locked again until unlock is complete."))
+	case lockWipe.IsPendingWipe():
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Host has pending wipe request. Cannot process lock requests once host is wiped."))
+	case lockWipe.IsLocked():
+		// TODO(mna): was not convered in figma, succeed quietly, returning the
+		// current unlock pin for macos? For now I'll return 409 conflict.
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Host is already locked.").WithStatus(http.StatusConflict))
+	}
+
+	// all good, go ahead with queuing the lock request.
+	return svc.enqueueLockHostRequest(ctx, host, lockWipe)
+}
+
+func (svc *Service) UnlockHost(ctx context.Context, hostID uint) (string, error) {
+	// First ensure the user has access to list hosts, then check the specific
+	// host once team_id is loaded.
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
+		return "", err
+	}
+	host, err := svc.ds.HostLite(ctx, hostID)
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "get host lite")
+	}
+
+	// Authorize again with team loaded now that we have the host's team_id.
+	// Authorize as "execute mdm_command", which is the correct access
+	// requirement.
+	if err := svc.authz.Authorize(ctx, fleet.MDMCommandAuthz{TeamID: host.TeamID}, fleet.ActionWrite); err != nil {
+		return "", err
+	}
+
+	// locking validations are based on the platform of the host
+	switch host.FleetPlatform() {
+	case "darwin":
+		// all good, no need to check if MDM enrolled, will validate later that it
+		// is currently locked.
+
+	case "windows", "linux":
+		// on windows and linux, a script is used to lock the host so scripts must
+		// be enabled
+		appCfg, err := svc.ds.AppConfig(ctx)
+		if err != nil {
+			return "", ctxerr.Wrap(ctx, err, "get app config")
+		}
+		if appCfg.ServerSettings.ScriptsDisabled {
+			return "", ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Can't unlock host because running scripts is disabled in organization settings."))
+		}
+
+	default:
+		// TODO(mna): should we allow/treat ChromeOS as Linux for this purpose?
+		return "", ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", fmt.Sprintf("Unsupported host platform: %s", host.Platform)))
+	}
+
+	lockWipe, err := svc.ds.GetHostLockWipeStatus(ctx, host.ID, host.FleetPlatform())
+	if err != nil {
+		return "", ctxerr.Wrap(ctx, err, "get host lock/wipe status")
+	}
+	switch {
+	case lockWipe.IsPendingLock():
+		return "", ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Host has pending lock request. Host cannot be unlocked until lock is complete."))
+	case lockWipe.IsPendingUnlock():
+		return "", ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("host_id", "Host has pending unlock request. The host will unlock when it comes online."))
```

Comment:
I think we need to special-case macOS here so the pin can be seen multiple times.